File: public/services/roadmap.ts
import { httpClient } from "@fider/services"
import { RoadmapData, RoadmapColumn } from "@fider/models"

export const roadmap = {
  async getRoadmap(): Promise<RoadmapData> {
    const response = await httpClient.get("/api/v1/roadmap")
    return response.data
  },

  async assignPostToColumn(postNumber: number, columnId: number, position: number): Promise<void> {
    await httpClient.post(`/api/v1/roadmap/posts/${postNumber}/assign`, {
      postNumber,
      columnId,
      position,
    })
  },

  async removePostFromRoadmap(postNumber: number): Promise<void> {
    await httpClient.delete(`/api/v1/roadmap/posts/${postNumber}/assign`)
  },

  async reorderPostInColumn(postNumber: number, newPosition: number): Promise<void> {
    await httpClient.put(`/api/v1/roadmap/posts/${postNumber}/position`, {
      postNumber,
      newPosition,
    })
  },

  async getColumns(): Promise<RoadmapColumn[]> {
    const response = await httpClient.get("/api/v1/admin/roadmap/columns")
    return response.data
  },

  async createColumn(name: string, isVisibleToPublic: boolean): Promise<RoadmapColumn> {
    const response = await httpClient.post("/api/v1/admin/roadmap/columns", {
      name,
      isVisibleToPublic,
    })
    return response.data
  },

  async updateColumn(id: number, name: string, isVisibleToPublic: boolean): Promise<RoadmapColumn> {
    const response = await httpClient.put(`/api/v1/admin/roadmap/columns/${id}`, {
      name,
      isVisibleToPublic,
    })
    return response.data
  },

  async deleteColumn(id: number): Promise<void> {
    await httpClient.delete(`/api/v1/admin/roadmap/columns/${id}`)
  },

  async reorderColumns(columnIds: number[]): Promise<void> {
    await httpClient.put("/api/v1/admin/roadmap/columns/reorder", columnIds)
  },
}

File: app/models/cmd/roadmap.go
package cmd

import (
	"github.com/getfider/fider/app/models/entity"
)

// AssignPostToColumn assigns a post to a roadmap column
type AssignPostToColumn struct {
	PostID       int
	ColumnID     int
	Position     int
	AssignedByID int
	Result       *entity.RoadmapAssignment
}

// RemovePostFromRoadmap removes a post from the roadmap
type RemovePostFromRoadmap struct {
	PostID       int
	TenantID     int
	RemovedByID  int
}

// ReorderPostInColumn changes the position of a post within its column
type ReorderPostInColumn struct {
	PostID       int
	NewPosition  int
	UpdatedByID  int
}

// CreateRoadmapColumn creates a new roadmap column
type CreateRoadmapColumn struct {
	TenantID          int
	Name              string
	Slug              string
	Position          int
	IsVisibleToPublic bool
	CreatedByID       int
	Result            *entity.RoadmapColumn
}

// UpdateRoadmapColumn updates an existing roadmap column
type UpdateRoadmapColumn struct {
	ColumnID          int
	Name              string
	IsVisibleToPublic bool
	UpdatedByID       int
	Result            *entity.RoadmapColumn
}

// DeleteRoadmapColumn deletes a roadmap column
type DeleteRoadmapColumn struct {
	ColumnID     int
	TenantID     int
	DeletedByID  int
}

// ReorderRoadmapColumns changes the order of roadmap columns
type ReorderRoadmapColumns struct {
	TenantID     int
	ColumnIDs    []int
	UpdatedByID  int
}

// GetMaxRoadmapColumnPosition gets the maximum position for roadmap columns
type GetMaxRoadmapColumnPosition struct {
	TenantID int
	Result   *int
}
